recipes-core/tdx-enc-handler/tdx-enc-handler/tdx-enc-caam.sh

```diff
@@ -61,6 +61,20 @@ tdx_enc_key_gen() {
     fi
 }
 
+# backup original data in partition (if not encrypted)
+tdx_enc_backup_data() {
+    mkdir -p "${TDX_ENC_STORAGE_MOUNTPOINT}"
+    if mount -t ext4 ${TDX_ENC_STORAGE_LOCATION} "${TDX_ENC_STORAGE_MOUNTPOINT}"; then
+        tdx_enc_log "Backing up original content..."
+        
+        mkdir /tmp/${TDX_ENC_DM_DEVICE}
+        chmod 777 /tmp/${TDX_ENC_DM_DEVICE}
+        mount -t tmpfs -o size=$(blockdev --getsz ${TDX_ENC_STORAGE_LOCATION}) ${TDX_ENC_DM_DEVICE} /tmp/${TDX_ENC_DM_DEVICE}
```

Comment:
Hmm... If we have 500GB of data in the partition, then we will use 500MB of RAM! That might not be desirable and can cause problems. I think we should set some limits here.

Maybe limit the ammount of data to be copied and failed if it passes a threshold? 

@rborn-tx any ideas?

Reply:
Is there an easy way to determine available RAM? if so I could add a check that this size is not more than 50-75% of available RAM and fail otherwise?
That way it would be generic for any given module.

Reply:
Interesting idea. Please limit to 30% of available memory.

This command will return free memory in kB:

```
# grep MemFree: /proc/meminfo | tr -s ' ' | cut -d ' ' -f 2
```

Reply:
That looks like a better approach indeed! Combined with making sure we don't fill up the RAM by limiting the maximum size of the tar file.

Reply:
> Both the BSP images and Torizon OS have /tmp as tmpfs, so my question is: why do we need this other tmpfs mount?
> 
> I notice the size of the tmpfs here is made to match the size of the whole block device whose data is to be saved which could be a waste, right? Perhaps a better approach would be to tar up the data to be saved into a simple file in /tmp. We could limit the size of the compressed file on-the-fly by doing something like this:
> 
> ```
> tar -C "${TDX_ENC_STORAGE_MOUNTPOINT}" -cj . | dd bs=1024 count=123 of=/tmp/saved.tar.bz2
> ```
> 
> where 123 is the size limit in kB (30% of the RAM as suggested).
> 
> With this, the amount of RAM temporarily used would be just the size of the data being saved after compression.
> 
> Wouldn't this work? Any downsides that you guys see?

I can certainly add the compressed tar.
With the tmpfs stuff.... (I copied code from somewhere else), I assume this is in RAM? but would it be mounted by the time this script executes?
I was also toying with the idea to make the percentage configurable (just in case) and could set the default to 30.

I will get these added to my local changes on the train home and give them a test on my image tomorrow and report back.

Reply:
> With the tmpfs stuff.... (I copied code from somewhere else), I assume this is in RAM?

It is.

> ...but would it be mounted by the time this script executes?

Good point, since the script is now touching `/tmp` (independent of the change I proposed) and we set `DefaultDependencies=no` I think the service should now explicitly require "tmp.mount" as a "Wants" dependency.

@sergioprado Do you agree?

Reply:
Initial testing looks good for these. Just looking at adding the tar bit and checking the tmpfs wants change. Should have a new commit on this later today.

Reply:
> Good point, since the script is now touching `/tmp` (independent of the change I proposed) and we set `DefaultDependencies=no` I think the service should now explicitly require "tmp.mount" as a "Wants" dependency.

I will add this to my local changes and see what happens. I assume this means I won't have to manually mount and create the tmpfs anymore too?

Reply:
One other question I have....
If this is now using the already mounted tempfs, do we need to have any error message/validation on the size of the backup?
I expect the tar commend mentioned would error if the size is greater than that?

Reply:
> Good point, since the script is now touching /tmp (independent of the change I proposed) and we set DefaultDependencies=no I think the service should now explicitly require "tmp.mount" as a "Wants" dependency. Do you agree?

Makes sense, though I would recommend checking systemd logs to make sure we don't have any cyclic dependencies.

> I assume this means I won't have to manually mount and create the tmpfs anymore too?

Exactly, /tmp should already be mounted with tmpfs. You don't have to mount it or unmounted it.

> If this is now using the already mounted tempfs, do we need to have any error message/validation on the size of the backup?

As a general rule, yes, show error messages before returning an error from the script.

> I expect the tar commend mentioned would error if the size is greater than that?

Yes!

Reply:
As it turns out, that command seems to fail for me. no option '-j' apparently. I have an alternative command that seems to work.
Also, I have a possibly better solution to check if the data is too big (assuming the tmpfs is already mounted and configured correctly).
Just running another test now and will push my changes if it works.

Reply:
@MMG-DG 

> As it turns out, that command seems to fail for me. no option '-j' apparently.

That's odd. If I run:

```
root@colibri-imx6-10866313:~# tar
BusyBox v1.35.0 () multi-call binary.

Usage: tar c|x|t [-ZzJjahmvokO] [-f TARFILE] [-C DIR] [-T FILE] [-X FILE] [LONGOPT]... [FILE]...
```

And a test seem to work.

> In my testing the command above didn't error.

I guess you're talking about the pipeline involving `dd`. That's true and reason is that when we have a shell pipeline like `a | b | c` the exit code is by default set to the last command in the pipeline; in our case the last command is `dd` which didn't fail but rather caused the failed to the preceding commands.

In bash we would solve this by simply doing `set -o pipefail` but since we're using `sh`, we need an alternative implementation such as:

```
msgs="$({ { tar -C "${TDX_ENC_STORAGE_MOUNTPOINT}" -cj . || echo "ERROR" >&2; } | dd bs=1024 count=100 of=/tmp/saved.tar.bz2; } 2>&1)"
if echo "${msgs}" | grep -q ERROR; then
    # Some error happened.
fi
```

If the tar `-j` option doesn't really work, the pipeline above becomes:
```
msgs="$({ { tar -C "${TDX_ENC_STORAGE_MOUNTPOINT}" -c . || echo "ERROR" >&2; } | { bzip2 -cz || echo "ERROR" >&2; } | dd bs=1024 count=100 of=/tmp/saved.tar.bz2; } 2>&1)"
```

Reply:
> If the tar `-j` option doesn't really work, the pipeline above becomes:

I will update my script and test locally on my dev board today and see what happens. I am wondering if there is a difference here because I am based on the tdx-reference-minimal-image? `tar` is available, `bzip2` and `gzip` are available, but tar still seems unable to use `-j` or `-z`.

Will do some more testing today and try to figure out the best option.

Reply:
From testing on my setup I get the following when using the `j` option:
![image](https://github.com/toradex/meta-toradex-security/assets/153825153/b2955614-1177-4558-bbbb-5c0d8ee18c8d)
Will revert to the separate bzip command and test again.

Reply:
I believe the only thing remaining from this thread is the ability to get all errors properly. Current code still have the check, but any time I try to include the `OR`'d pipe for errors it fails.
Any ideas welcome.
I was debating about having the commands push the stderr into a file in `/tmp` and then have the check read that file?

Reply:
@MMG-DG Can you update the PR so we can look at the actual code?

Reply:
@rborn-tx - I am looking at the PR and all the code is here. Not sure why you don't see it.
I have pushed all changes to my fork's branch and it shows up here.
My guess is that because the lines have changes that this initial comment was on is not out of date it still points to the old code.

Reply:
@sergioprado & @rborn-tx  - Would you both be happy to resolve this comment as the latest code has all these changes in and if needed start a new conversation based on the latest code if needed?